src/ProtoRoute/Message.hs

```diff
@@ -41,6 +54,33 @@ fromTV (Opt txt)  = show txt
 fromTV (Rep list) = show list
 fromTV Error = "Error"
 
+checkFields :: (FieldValue, (OnlyFRule, OnlyFType)) -> Bool
+checkFields ((FText tvA), (rA, fTA)) = fTA == Sstr && (correct tvA rA)
+checkFields ((FInt tvB), (rB, fTB))  = fTB == Number && (correct tvB rB)
+checkFields ((FMsg tvC), (rC, fTC))  = fTC == Msg && (correct tvC rC)
+
+correct :: (Eq a) => TValue a -> OnlyFRule -> Bool
+correct (Req _) TReq = True
+correct (Opt _) TOpt = True
+correct (Rep _) TRep = True
+correct _ _ = False
+
+-- Not quire sure how to check for empty required fields here
+validateMessage :: Message -> MessageSchema -> Either ValidationError ()
```

Comment:
It's better to write function in recursive way. So now you have:

```haskell
validateMessage :: Message -> MessageSchema -> Either ValidationError ()
```

After that you can introduce:

```haskell
validateField :: MessageField -> SchemaField -> Either ValidationError ()
```

At this point you might notice that it also makes sense to introduce data type like:

```haskell
data MessageField = MessageField
    { messageFieldName :: FieldName
    , messageFieldValue :: FieldValue
    }
```